### ISSUE TITLE
It should probably allow you to do `spread` on an arbitrary promise?

### ISSUE THREAD

**Issue by Kuzmin** [showpad/angular-q-spread#5]
I want to be able to do something like 

    promise1.then(function () {
        return $q.all([promise2, promise3]);
    }).spread(function () { ... });

but right now the .spread method is only added to the return value of $q.all. I think adding the method to the Promise prototype instead of only adding it to the return value of $q.all would do the trick!

**Comment by devinivy** [comment]
+1 definitely!

**Comment by stevematyas** [comment]
Similar to #7.

**Comment by klaascuvelier** [comment]
@Kuzmin, @devinivym @stevematyas:
Sadly it is not possible to add the spread method on promises by default as the Promise object is private. See: https://github.com/angular/angular.js/blob/master/src/ng/q.js#L271-297 and https://github.com/angular/angular.js/blob/master/src/ng/q.js#L578-582

It would be possible by replacing the whole `$delegate` service, but I did not want to copy past angular source code, I only wanted to decorate methods.

I was able to add the `spread` method to more `$q` methods: `when`, `all`, `resolve`, `defer`.
I also decorated the `then` methods on promises created through on of the above decorated methods to support `spread` as well and made sure this decorator was called recursively on all then methods.

I wrote a test which resembles your example, @Kuzmin.

Can you guys review PR? https://github.com/showpad/angular-q-spread/pull/10

**Comment by kasperlewau** [comment]
@klaascuvelier 

Out of curiosity I attempted to apply something similar to what I did in [angular-q-thenresolve](https://github.com/kasperlewau/angular-q-thenresolve/blob/master/src/angular-q-thenresolve.js) to resolve the proposed issue. 

I based it off of #10, left the spec suite **as is** and replaced the decorator with the following:

```js
$provide.decorator('$q', ['$delegate', function ($delegate) {
  var originalDefer = $delegate.defer;

  $delegate.defer = function () {
    var promiseProto = originalDefer().promise.constructor.prototype;

    Object.defineProperty(promiseProto, 'spread', {
      value: function (resolve, reject) {
        function spread (data) {
          return resolve.apply(void 0, data);
        }

        return this.then(spread, reject);
      },
      writable: true,
      enumerable: false
    });

    return originalDefer();
  }

  return $delegate;
}]);
```

I haven't tested this in the wild, but it does run through the spec suite of #10 with flying colours. 

So you could actually expose it on the promise prototype. And you're not really overwriting $q, so much as using it as a gateway to do some grunt work on the promise prototype.

---

*In saying that, it may not even be needed to dig down on the prototype level with the changes you've introduced. Changes look good the way they are imho. I only ever wanted to highlight an alternative route of introducing the `.spread` method across the board.*

**Comment by klaascuvelier** [comment]
@kasperlewau thanks man! Your method looks much cleaner.
I did not know there was a way to access the prototype of the promise this way. Great.
I created this new PR and I am going to merge in that one pretty soon: https://github.com/showpad/angular-q-spread/pull/11

**Comment by kasperlewau** [comment]
I'm glad I could be of some assistance - you are most welcome! :smile:

**Comment by klaascuvelier** [comment]
fixed in #11
